### PR TITLE
FFmpeg probe fps

### PR DESCRIPTION
**Pull request by smellslikeml**
According to [YouTube](https://support.google.com/youtube/answer/1722171?hl=en#zippy=%2Cframe-rate), common frame rates include: 24, 25, 30, 48, 50, 60 frames per second but some applications requiring specialized cameras capture at hundreds or thousands of frames per second.

Relating frame numbers and time-scale in video requires a conversion.

This udf uses ffmpeg [probe](https://kkroening.github.io/ffmpeg-python/#ffmpeg.probe) to query average frame rate from a video's metadata.

### CODE REVIEW

File: python/rikai/spark/functions/vision.py [hunk 1]
```diff
@@ -43,6 +43,7 @@
     "image_copy",
     "numpy_to_image",
     "video_to_images",
+    "video_framerate",
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/functions/vision.py:46]: Btw, is it possible to name this function `framerate` instead? I am not sure what chance `framerate` can be applied to other data types. But lets claim the shortest name when we have a chance?

File: python/rikai/spark/functions/vision.py [hunk 2]
```diff
@@ -170,6 +171,47 @@ def video_to_images(
...
+    assert isinstance(
+        video, (YouTubeVideo, VideoStream)
+    ), "Input type must be YouTubeVideo or VideoStream"
+    video_uri = (
+        video.get_stream().uri
+        if isinstance(video, YouTubeVideo)
+        else video.uri
+    )
+
+    n, d = map(
```
[the remark this answers is not in the document]
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:207]: ffmpeg probe returns a dictionary,  the 'avg_frame_rate' key maps to a string like: '30000/1001' to indicate 30 fps.

File: python/rikai/spark/functions/vision.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:46]: `framerate` sounds good 👍 

Looking into [opencv](https://opencv-java-tutorials.readthedocs.io/en/latest/) and [ffmpeg](https://github.com/bramp/ffmpeg-cli-wrapper) java bindings to implement in `spark.functions.video`.

File: python/rikai/spark/functions/vision.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/spark/functions/vision.py:207]: does it mean that `d` wont be empty or zero?

[the remark this answers is not in the document]
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:207]: I dont think so but I can borrow from this [example](https://github.com/kkroening/ffmpeg-python/blob/4cb7d26f55208b01b123f07b13108bf4143794e2/examples/video_info.py)

File: python/rikai/spark/functions/vision.py [hunk 3]
```diff
@@ -170,6 +171,47 @@ def video_to_images(
...
+        raise ValueError(
+            "Couldn't import ffmpeg. Please make sure to "
+            "`pip install ffmpeg-python` explicitly or install "
+            "the correct extras like `pip install rikai[all]`"
+        )
+    assert isinstance(
+        video, (YouTubeVideo, VideoStream)
+    ), "Input type must be YouTubeVideo or VideoStream"
+    video_uri = (
+        video.get_stream().uri
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/rikai/spark/functions/vision.py:202]: base class: youtubevideo.get_stream() just gets some defaults. I'm wondering if `YouTubeVideo.uri` will "hide" too much complexity. In addition, will people think the uri is the youtube url OR the actual uri of the video?

Though I do see the benefit of simplicity. How do you think we can improve the API here?

File: python/rikai/spark/functions/vision.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (smellslikeml) [reply python/rikai/spark/functions/vision.py:207]: can expand to extract `number of frames` and `duration`:

image

as well as `height`, `width`, other stream info.

and

**Comment by changhiskhan** [comment]
does #177 replace this PR or?

**Comment by smellslikeml** [comment]
> does #177 replace this PR or?

I think so, #177 will help probe additional info in the context of loading video.

**Pull request closed without merging**